File: SpringBootCache/src/test/java/com/example/SpringBootCacheApplicationTests.java
package com.example;

import com.example.component.BookRepository;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.test.context.junit4.SpringRunner;

import com.example.domain.Book;

@RunWith(SpringRunner.class)
@SpringBootTest
public class SpringBootCacheApplicationTests {

	@Autowired
	private BookRepository repository;

	private long startTime;
	private long endTime;
	
	private static final Logger logger = LoggerFactory.getLogger(SpringBootCacheApplicationTests.class);
	
	@Before
	public void onBefore() {
		startTime = System.currentTimeMillis();
	}
	
	@After
	public void onAfter() {
		endTime = System.currentTimeMillis();
		logger.info("소요시간: {}ms", endTime-startTime);
	}
	
	@Test
	public void test1() {
		repository.getByIsbn("a");
	}

	@Test
	public void test2() {
		repository.getByIsbn("a");
	}
	

	@Test
	public void test3() {
		repository.getByIsbn("b");
	}
	
	@Test
	public void test4() {
		repository.getByIsbn("a");
	}
	
	@Test
	public void test5() {
		repository.refresh("a");
		repository.getByIsbn("a");
	}
}

File: SpringBootCache/src/test/kotlin/com/example/kotlin/SpringBootCacheApplicationTests.kt
package com.example.kotlin

import com.example.kotlin.component.BookRepository
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.cache.CacheManager

@SpringBootTest
class SpringBootCacheApplicationTests {

    @Autowired
    private lateinit var repository: BookRepository
    @Autowired
    private lateinit var cacheManager: CacheManager

    @Test
    fun `book 캐시에 a에 해당하는 값이 없다`() {
        cacheManager.getCache("book")?.let {
            assertThat(it.get("a")).isNull()
        }
    }

    @Test
    fun `b를 저장하면 book 캐시에 a에 해당하는 값이 있다`() {
        cacheManager.getCache("book")?.let {
            repository.getByIsbn("b")
            assertThat(it.get("b")).isNotNull
        }
    }

    @Test
    fun `c를 저장하면 book 캐시에 값이 있지만 refresh하면 값이 없다`() {
        cacheManager.getCache("book")?.let {
            repository.getByIsbn("c")
            assertThat(it.get("c")).isNotNull
            repository.refresh("c")
            assertThat(it.get("c")).isNull()
        }
    }
}
